tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +182,215 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
+                    status.notify_all();
+                    return false;
+                }
+                break;
+            case AllocationStatus::ALLOCATING:
+            case AllocationStatus::ALLOCATED:
+                // Allocation already started, will have to deallocate
+                if (status.compare_exchange_weak(current_status, AllocationStatus::DEALLOCATION_REQUESTED)) {
+                    status.notify_all();
+                    return true;
+                }
+                break;
+            case AllocationStatus::DEALLOCATION_REQUESTED:
+            case AllocationStatus::DEALLOCATED:
+                // Deallocation was already started, nothing to be done
+                return false;
         }
     }
-    return *this;
-}
-
-Buffer::Buffer(Buffer &&other) :
-    device_(other.device_),
-    size_(other.size_),
-    address_(other.address_),
-    page_size_(other.page_size_),
-    buffer_type_(other.buffer_type_),
-    buffer_layout_(other.buffer_layout_),
-    shard_parameters_(std::move(other.shard_parameters_)),
-    bottom_up_(other.bottom_up_),
-    buffer_page_mapping_(std::move(other.buffer_page_mapping_)),
-    allocate_(other.allocate_) {
-    // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-    // transferred to `this`
-    other.device_ = nullptr;
-}
-
-Buffer &Buffer::operator=(Buffer &&other) {
-    if (this != &other) {
-        this->device_ = other.device_;
-        this->size_ = other.size_;
-        this->address_ = other.address_;
-        this->page_size_ = other.page_size_;
-        this->buffer_type_ = other.buffer_type_;
-        this->buffer_layout_ = other.buffer_layout_;
-        this->shard_parameters_ = std::move(other.shard_parameters_);
-        this->bottom_up_ = other.bottom_up_;
-        this->buffer_page_mapping_ = std::move(other.buffer_page_mapping_);
-        this->allocate_ = other.allocate_;
-        // Set `other.device_` to be nullptr so destroying other does not deallocate reserved address space that is
-        // transferred to `this`
-        other.device_ = nullptr;
+}
+
+void Buffer::deallocate() {
+    if (!prepare_deallocation(allocation_status_)) {
+        return;
+    }
+
+    device_->push_work([self = weak_self.lock()] {
+        if (self->device_->initialized_ && self->size_ != 0) {
+            detail::BUFFER_MAP.erase({self->device_->id(), self->address_});
+            detail::DeallocateBuffer(self.get());
+        }
+
+        self->allocation_status_ = AllocationStatus::DEALLOCATED;
+    });
+}
+
+void Buffer::deleter(Buffer* buffer) {
+    // There is no concurrent allocations/deallocations happening, so no extra checks are required
+    if (buffer->allocation_status_ == AllocationStatus::DEALLOCATED) {
+        return;
+    }
+
+    buffer->device_->push_work([buffer] {
+        std::unique_ptr<Buffer> unique_buffer = std::unique_ptr<Buffer>(buffer);
```

Comment:
What's the reason for creating a `unique_buffer` here?

Reply:
To deallocate the call `delete buffer` at the scope exit